### PR TITLE
Add support for `@google-cloud/cloud-sql-connector` database connection

### PR DESCRIPTION
**Pull request by benjdlambert**
- chore: support cloudsql-connector
- chore: add dev dep and changeset Signed-off-by: blam <ben@blam.sh>

#### :heavy_check_mark: Checklist

- [ ] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [ ] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: packages/backend-defaults/src/entrypoints/database/connectors/postgres.ts [hunk 1]
```diff
@@ -64,18 +64,45 @@ export function createPgDatabaseClient(
...
   );
+
+  if (config.client === 'pg+google-cloudsql') {
+    const {
+      Connector: CloudSqlConnector,
+      IpAddressTypes,
+      AuthTypes,
+    } = await import('@google-cloud/cloud-sql-connector');
+    // override the config to be pg for backwards compat with other code
+    config.client = 'pg';
```
Comment (freben) [inline packages/backend-defaults/src/entrypoints/database/connectors/postgres.ts:87]: is the result of `config.get()` safely mutable? 🤔 Maybe you could add `{}` as the first argument to `mergeDatabaseConfig` to make sure that it's a separate object?

File: packages/backend-defaults/src/entrypoints/database/connectors/postgres.ts [hunk 2]
```diff
@@ -64,18 +64,45 @@ export function createPgDatabaseClient(
...
+
+    const connector = new CloudSqlConnector();
+    const clientOpts = await connector.getOptions({
+      instanceConnectionName: dbConfig.getString('instanceConnectionName'),
```
Comment (freben) [inline packages/backend-defaults/src/entrypoints/database/connectors/postgres.ts:91]: note that this and the above new client string should have a corresponding `config.d.ts` [update](https://github.com/backstage/backstage/blob/03351abbddbc5bbee85afcdcbc99517af199c7ee/packages/backend-defaults/config.d.ts#L380). Actually, the instance connection name (and possibly the ip and auth types etc) should probably go in the connection object in the future.

```yaml
client: 'pg+google-cloudsql'
connection: 'the-cloudsql-connection-identifier-thing'
```

as a shorthand and then

```yaml
client: 'pg+google-cloudsql'
connection:
  instance: 'the-cloudsql-connection-identifier-thing'
  ip: public
  ...
```

as the long form when we add support for more parameters

**Review by freben — APPROVED** [review]
Nice!

File: packages/backend-defaults/src/entrypoints/database/connectors/postgres.ts [hunk 3]
```diff
@@ -64,18 +64,53 @@ export function createPgDatabaseClient(
...
+    const connector = new CloudSqlConnector();
+    const clientOpts = await connector.getOptions({
+      instanceConnectionName: config.connection.instance,
+      ipType: IpAddressTypes.PUBLIC,
+      authType: AuthTypes.IAM,
+    });
+
+    const cloudsqlConfig = {
+      ...config,
+      client: 'pg',
```
Comment (Rugvip) [inline packages/backend-defaults/src/entrypoints/database/connectors/postgres.ts:99]: Thinking we should throw if `client` isn't set to `pg` too

**Review by Rugvip — APPROVED** [review]
Yuuuuuuuuup :shipit: 

<sub><sub>just a minor thing</sub></sub>

**Pull request merged**